addons/pkg-npm/jaxrs/src/main/java/org/commonjava/indy/pkg/npm/jaxrs/NPMContentAccessHandler.java

```diff
@@ -394,31 +394,28 @@ private void generateNPMContentsFromTransfer( final Transfer transfer, final Eve
 
             return;
         }
-        try
-        {
-            if ( target1 == null || target2 == null )
-            {
-                return;
-            }
-            out1 = target1.openOutputStream( TransferOperation.UPLOAD, true, eventMetadata );
-            out1.write( content1.getBytes() );
 
-            out2 = target2.openOutputStream( TransferOperation.UPLOAD, true, eventMetadata );
-            out2.write( Base64.decodeBase64( content2 ) );
+        if ( versionTarget == null || tarballTarget == null )
+        {
+            return;
+        }
 
-            setGeneratedTransfers( transfer, target1, target2 );
+        try (OutputStream versionOutputStream = versionTarget.openOutputStream( TransferOperation.UPLOAD, true,
+                                                                                eventMetadata );
+             OutputStream tarballOutputStream = tarballTarget.openOutputStream( TransferOperation.UPLOAD, true,
+                                                                                eventMetadata ))
+        {
+            versionOutputStream.write( versionContent.getBytes() );
+            tarballOutputStream.write( Base64.decodeBase64( tarballContent ) );
+            generatedTransfers = new GeneratedTransfers( transfer, versionTarget, tarballTarget );
```

Comment:
As near as I can tell, this is reassigning the value of the field `generatedTransfers`. What happens if two NPM uploads happen at the same time? Will there be a race to determine which one sets the value of `generatedTransfers` and cross-pollutes the execution of the other one?

What I was hoping was to make this method (`generateNPMContentsFromTransfer()`) *return* a GeneratedTransfers instance back to `doCreate()`, where it's saved as a variable. Then, pass this variable in as a *parameter* to the `generateHttpMetadataHeaders()` method that needs it. That will keep the state scoped to the execution and avoid side effects. It will also avoid any need to clean up a component-level field.

Reply:
@jdcasey, I use map to trace the target key need to be generated for per call, while it may indeed cause the issue when multi uploads for one project and different version, may meet the missing of some version/tarball transfers storing as a result of race condition.

So I change this into a private method, which is only need to store simply in a list instead of private class I think.